xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xpn.xwiki.internal.doc;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.ListIterator;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.xpn.xwiki.doc.XWikiAttachment;
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @since 10.0RC1
+ */
+public abstract class XWikiAttachmentList extends ArrayList<XWikiAttachment>
+{
+
+    private Map<String, XWikiAttachment> map;
+
+    /**
+     * Initializes the map.
+     * 
+     * @since 10.0RC1
+     */
+    public XWikiAttachmentList()
+    {
+        map = new TreeMap<String, XWikiAttachment>();
```

Comment:
I did a quick check and TreeMap is not thread safe so you will have issues with it (XWikiDocument can be manipulated from several threads at the same time). ConcurrentSkipListMap seems be here for that.

By the way you don't really need this XWikiAttachmentList constructor, you could directly initialize `map` field.

Reply:
now since I added the document parameter, seems like better practice to initialize it in the constructor

Reply:
okay sure, I can move it and make it final